Implement the RelationRegistry as a graph

snusnu opened the pull request:
Opening a new PR for this feature as the old one included some unrelated commits that were better off in master for now.

[Here](http://postimage.org/image/fe3m1zjox/) are some sketches of how I imagine the graph to look like when performing some common operations.

For reference, [this](https://github.com/solnic/dm-mapper/pull/8) was the previous PR.

snusnu committed:
Initial code implementing RelationRegistry as a graph

snusnu committed:
Create edges and nodes for single hop relationships

snusnu committed:
Move guard from method to caller

snusnu committed:
Introduce a local variable for better formatting

snusnu commented:
@dkubb is f3f5fe6a064c578121cb10976f3a48381278f511 going into the direction you've imagined?

snusnu committed:
Make aliasing info available in Node::Relation objects

snusnu committed:
Generate mappers based on relation registry info

snusnu commented:
@solnic @dkubb @mbj All relationship mappers are now generated from the information in the graph and it works!?! The code is still messy but we can refactor from here.

Currently there are no specs for `has 1, :thing, :through => :something` and none for "going through through" in general. Maybe they would work, maybe not, no idea.

Any thoughts much appreciated!

snusnu committed:
Nuke MapperRegistry#clear which was only needed for specs

solnic committed:
Merge branch 'master' into relation_registry_graph

Conflicts:
	lib/dm-mapper.rb

solnic committed:
Add pending spec for 1:1-through

solnic commented:
@snusnu I looked at the code in the morning but I didn't have enough time to deeply dive into it. I merged master in and added a pending spec for "1:1 through", pending as it fails - see my FIXME comment in `AttributeSet#remap`.

Overall I think this is a good direction. With such clear representation of relationships it's easier to deal with aliasing, dependencies and all that crap. The code needs a serious clean up though. Maybe it's time to end the spike phase and focus on unit tests and refactor? This will help in cleaning up stuff and actually finishing the implementation. I can help with that :)

Let's discuss further in #datamapper channel :)

snusnu committed:
Fix relationship_one_to_one_through_spec

snusnu committed:
Add pending relationship_many_to_many_through_spec

mbj commented:
@snusnu I through about the problem of building a graph with immutable nodes and edges. In the current design we reference the edges of a node and vice versa via instance variables, creating a direct transitive dependency. When nodes and edges would be immutable we would have an chicken egg problem.

As the current code must mutate the nodes and edges when new nodes and edges are added our design is more complex. See the unclear CQS role of `EdgeSet#add` as a symptom, we do have to care about object identity and not equality here. In summary we suffer from direct transitive referencing, with all its problems.

How about the following design that would break the explicit and transitive strong referencing between edges and nodes:

``` ruby
# Missing the Connector in this example but hopefully it is able to communicate the idea
#
# Also none of the relation registry specific stuff is not implemented. 
# The node name symbolizes the payload.
# Edge is missing payload.
#
# The idea is NOT to store foreign objects inside the nodes. 
# Thus breaking the direct transitive referencing.
# As an example edges for a node are loaded indirectly via the graph object.
#
# The graph is mutable in this example, a more advanced implementation could s
# tore the graph within the relationship registry and update the graph in a more functional 
# style (storing mutated copy).
#
class Graph
  # Return edges
  #
  # @return [Enumerable<Edge>]
  #
  # @api private
  #
  attr_reader :edges

  # Return nodes
  #
  # @return [Enumerable<Node>]
  #
  # @api private
  #
  attr_reader :node

  # Return edges for node
  #
  # @param [Node] node
  #
  # @return [Enumerable<Edge>]
  #
  # @api private
  #
  def edges_for(node)
    @edge_cache[node] ||= edges.select { |edge| edge.connects?(node) }
  end

  # Add node with name to graph
  #
  # @param [Symbol] name
  #
  # @return [Enumerable<Edge>]
  #
  # @api private
  #
  def node(name)
    nodes << Node.new(self, name)
    reset
    self
  end

  # Add node to graph
  #
  # @param [Symbol] node
  #
  # @return [self]
  #
  def edge(edge)
    edges << edge
    reset
    self
  end

private

  # Initialize object
  #
  # @return [undefined]
  #
  # @api private
  #
  def initialize
    @edges, @nodes = Set.new, Set.new
    reset
  end

  # Reset caches
  #
  # @return [undefined]
  #
  # @api private
  #
  def reset
    @edge_cache = {}
  end

end

class Graph

  # An edge that connects two nodes left and right
  #
  # The name attribute is the payload that should be substituted with stuff we need for the relationship graph
  #
  class Edge
    include Immutable, Equalizer.new(:name, :left, :right)

    # Return name of node
    # 
    # @return [Node]
    #
    # @api private
    #
    attr_reader :name

    # Return left node
    # 
    # @return [Node]
    #
    # @api private
    #
    attr_reader :left

    # Return right node
    # 
    # @return [Node]
    #
    # @api private
    #
    attr_reader :right

    # Test if edge connects node
    #
    # @param [Node] node
    #
    # @return [true]
    #   if node is connected to edge
    #
    # @reutrn [false]
    #   otherwise
    #
    # @api private
    #
    def connects?(node)
      left == node || right == node
    end

    # Return name of node
    #
    # @return [Graph] graph
    #   the graph node belongs to
    #
    # @return [Symbol] name
    #   the name of the node
    #
    # @api private
    #
    def initialize(name, left, right)
      @name, @left, @right = graph, left, right
    end
  end

  # Node within a graph
  #
  class Node
    # I do not know the deduplication semantics of node so I just think it is deduped based on payload (name here).
    include Immutable, Equalizer.new(:name)

    # Return graph node belongs to
    #
    # @return [Graph]
    #
    # @api private
    #
    attr_reader :graph

    # Return name of node
    #
    # @return [Symbol]
    #
    # @api private
    #
    attr_reader :name

    # Return edges
    #
    # Delegates the edges to graph breaking transitive referencing,
    # and removing the need of taking care of identity.
    #
    # @return [Enumerable<Edge>]
    #
    # @api private
    #
    def edges
      graph.edges_for(self)
    end

  private

    # Initialize object
    #
    # @return [Graph] graph
    #   the graph node belongs to
    #
    # @return [Symbol] name
    #   the name of the node
    #
    # @api private
    #
    def initialize(graph, name)
      @graph, @name = graph, name
    end

  end
end
```

snusnu committed:
Move code closer to the relevant information

snusnu committed:
Fix intermediary relationship mapping for m:n

Previously, the generated intermediate relationship
was overwriting the original underlying 1:n used
to "go through".

snusnu committed:
[WIP] Provide shorter column aliases

The plan is to refactor the aliasing code soon, so
that it can handle infinite through relationships.

snusnu committed:
[WIP] Support going through a through relationship

This currently only works when going through a
through relationship, where that 2nd through
relationship does *not* go through anything.

What this means is, infinite through is not yet
supported.

snusnu committed:
Add pending spec for going through through

snusnu committed:
Construct only one node per added relationship edge

snusnu committed:
Nuke dead codes

dkubb commented:
@snusnu sorry for not responding. I've been tied up with work and back-to-school (not for me, but for my kids). I'm finally getting caught up on things and should be able to look at this shortly.

solnic committed:
Add simple graph implementation to support libs

solnic committed:
Simplify a little RelationRegistry by inheriting from Graph class

solnic committed:
[WIP] Simplify relation registry graph

This is a rather huge change and not completed at all but it looks like
even very complex multi-through joins work (on relation level).

Here's a sum up of those changes:

  * RelationRegistry inherits from base Graph class it holds edges
    (RelationConnector) and nodes (RelationNode)

  * RelationNode has name and an alias set needed for joins

  * RelationConnector has a relationship and left and right relation
    nodes that are used for join

  * RelationConnector::Builder is used to create nodes and connectors
    (edges) - if a relationship is created via other relationship then
    the builder creates a new connector and registers its relation as
    normal node so it can be used later for other connectors

  * RelationConnector::Builder creates "via" connectors recursively so
    it can go deep into multiple-through relationships and create
    correct joined relations with correct aliasing

This is a total mess right now but I think the hardest problem is
solved.

What's left to do:

  * Finish mapping part for joined and aliased relations

  * Create a nice interface for RelationRegistry so it's easy to walk
    through the graph when needed

  * Refactor code and add unit tests

ps. I left previous implementation in data_mapper/__relation_registry/
directory for reference

Note that this is heavily based on @snusnu's work and @mbj's suggestion
about simpler graph representation.

Note about aliasing:

Aliasing is simplfied here - there's no need for building unique aliases
for all nodes...and even if there is we can add it later and use this
"basic" implementation as the foundation. We simply prefix all the
fields with the singular relation name (like user_name, song_title
etc.).

solnic commented:
@dkubb I noticed that veritas sql generator creates an extra nested selected every time you do a Relation#rename so we end up with queries like that:

``` sql
SELECT "song_id", "song_title", "tag_id", "tag_name" FROM (
  SELECT "song_id", "song_title", "tag_id" FROM (
    SELECT * FROM (
      SELECT "id" AS "song_id", "title" AS "song_title" FROM "songs"
    ) AS "left"
    NATURAL JOIN (
      SELECT "song_id", "tag_id" FROM "song_tags"
    ) AS "right") AS "songs_song_tags") AS "left"
    NATURAL JOIN (
       SELECT "id" AS "tag_id", "name" AS "tag_name" FROM "tags"
    ) AS "right"
```

Notice the third SELECT. Is this redundant or am I missing something?

Do you think it should be easy to optimize the relation somehow or tweak the generator to produce a better query?

solnic committed:
Fix relation registry specs

solnic committed:
Fix alias set spec

solnic committed:
Lock down rspec to 1.3.1 again

solnic committed:
Fix unit tests

solnic committed:
Add spec for RelationRegistry#add_node

solnic committed:
Point to veritas-do-adapter on github again

solnic committed:
Add AliasSet#merge and AttributeSet#merge to simplify aliasing in connectors and nodes

solnic committed:
Add tests for Mapper::Builder#mapper

solnic committed:
Make multi-through mapping work FUCK YEAH :mega: :fire: :bomb:

snusnu committed:
Fix constant lookup in integration spec helper

snusnu committed:
Make AliasSet mutable (at least for now)

Adamantium is doing too much freezing here. Making
AliasSet immutable results in "can't modify frozen
object" during EV/EC attribute mapper finalization

snusnu committed:
Workaround RelationNode not quacking like a relation

It might be nice to have RelationNode quack like a
regular Relation at some point, but we've decided
to only support that once a more concrete usecase
arises.

snusnu committed:
Workaround RelationNode not quacking like Relation

solnic committed:
Fix spec for 1.8

solnic committed:
Remove obsolete TODO comments

solnic committed:
Make m:m through with an extra operation work

snusnu committed:
Fix spec for 1:n with additional restriction

snusnu committed:
Re-enable previously pending spec as it passes now

snusnu committed:
Add spec for restricting multi-through relationships

snusnu committed:
Rename a local variable in m2m multi-through spec

snusnu committed:
Add spec for restricting 1:1 through relationships

snusnu committed:
Nuke dead codes

solnic committed:
Make 1:1-through work

solnic committed:
Fix unit specs

snusnu committed:
Change unit specs to 1.8.7 compliant syntax

snusnu committed:
Mark m2m integration spec as pending for 1.8.7

solnic committed:
Deal with relationship aliasing on the connector side

solnic committed:
Simplify relation registry and add more tests

solnic committed:
Add tests for RelationConnect::Builder and via relationships

solnic committed:
Store nodes prepared for via joins in the graph as well

solnic committed:
Add RelationRegistry#build_node and #build_edge

solnic committed:
Remove dead code

solnic committed:
Enable specs that now pass on 1.8.7

solnic committed:
Use existing edges instead of creating duplicates

snusnu committed:
Enable specs that now pass on 1.8.7

snusnu committed:
Remove all constants before(:all) integration specs

solnic committed:
Refactor finalizer and make it finalize relation graph first

snusnu committed:
Nuke spiky codes

solnic committed:
Make RelationNode quack like a veritas relation

snusnu committed:
Push Mapper.aliases down to Mapper::Relation::Base.aliases

I have a feeling that eventually, this will be
delegated to the underlying RelationNode. Once
that's done, Mapper::Relation could maybe also
respond to this method, if it's even necessary.

snusnu committed:
Rename Mapper.relation_gateway to Mapper.gateway_relation

snusnu committed:
Don't default to Veritas::Relation::Empty

This just isn't useful. The usecase to be supported
here is allowing to define custom relations for
mappers. This is still possible by overwriting
Mapper::Relation.relation, however, returning nil
by default seems to be a better option.

snusnu committed:
Clear {mapper,relation}_registry before(:each) spec

snusnu committed:
Only finalize base relation mappers on first pass

snusnu committed:
Always give mocked (base relation) mappers a name

snusnu committed:
Rename to Finalizer#finalize_base_relation_mappers

snusnu committed:
Define methods in the order they are being called

snusnu commented:
That's probably even better, yeah.

snusnu committed:
Raise NotImplementedError instead of returning nil

snusnu committed:
Rename #for_relationship to #aliased_for

snusnu committed:
Rename #aliases_for_relationship to #aliases_for

snusnu committed:
Remove currently unused AliasSet#empty?

snusnu committed:
Add specs for AliasSet#exclude

snusnu committed:
Add specs for AliasSet#each

snusnu committed:
Start using the extracted inflector gem

Currently, inflection rules are still hosted in
dm-mapper itself, as the inflector gem doesn't
yet include them.

snusnu committed:
Nuke dead spec_helper codes

snusnu commented:
@solnic Just to give you some food for thought, do you notice how the current graph doesn't really make a distinction between edges and "connector" objects anymore? Since currently an edge "has a" relationship, it obviously isn't the bidirectional link @dkubb was talking about initially (see https://github.com/solnic/dm-mapper/pull/6 ).

Also, I don't see how the system registers a new node for a new edge. Since an edge basically produces a new relation, joining 2 others together, surely a new node for that new relation should be generated. Instead, the 2 "sides" of the join are registered as nodes, although that actually only produces a new node in the case of `through` relationships (as shown by the fact that all other branches in the current edge builder simply lookup existing nodes).

Altogether we end up not having separated the edge (method of combining 2 relations) from the name/method of access on the objects (the named relationships). In the system as it is now, it's also impossible to have the same edge be used by 2 relationships on both sides, again due to edge having a reference to the relationship.

Now, you know me, no pun intended, but it feels like quite some concepts got abandoned for the jersey thing (for non-insiders: aliasing in multi-through scenarios). Sure the code is simpler now (and I know damn well that we need the jersey thing), but it feels like quite some previously existing concepts need to be baked back in?

I might very well miss something important here, and also I'm perfectly ok with adding stuff back in, if **and only if** it makes sense to do so. I just wanted to get this out, so that we can discuss further! I'm also linking my initial diagram sketch [again](http://postimage.org/image/fe3m1zjox/) as I still think that it captures @dkubb's initial idea nicely and shows how the parts fit together.

I realize that the current code is able to pass more specs than the previous one, and that it's conceptually simpler. I also realize that it's not meant to be complete at this point. At the same time, I have the feeling that it doesn't fully represent the initial graph idea anymore. I'm not saying that's bad, I just want to be clear about the implications, and what it means for us moving forward.

dkubb commented on lib/data_mapper/mapper_registry.rb:
```diff
@@ -2,20 +2,60 @@ module DataMapper
 
   class MapperRegistry
 
+    class Identifier
```
This seems to do pretty much the exact same thing as `Equalizer`. Why not make sure the model and relationship objects use it, and you should get this behavior mostly for free?

EDIT: If you feel like you still need this object in order to create a "key" inside the mappers, you could just use `Equalizer` here and cut out half the code from this class besides the attributes and parts of the constructor.

solnic replied:
I tried to use equalizer but since it uses adamantium too many things become frozen and the code breaks badly. I will try again though.

mbj replied:
@solnic Try to use the do `no-deep-freeze-etc` branch of adamantium, IMHO it should work nicely.

dkubb commented on lib/data_mapper/mapper_registry.rb:
```diff
@@ -2,20 +2,60 @@ module DataMapper
+
     # @api public
     def initialize(mappers = {})
       @_mappers = mappers
```
fwiw, I don't think ivars within an object need to have leading underscores. Pretty much the only time I use this is when I'm adding something from an external source to an object; but something that I do not want the object's methods to use directly. In this case the list of mappers is part of the registry state, so I think it should be named how you'd expect.

dkubb commented on lib/data_mapper/mapper/builder.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+
+      def mapper_name
+        "#{@source_model.name}_X_#{Inflector.camelize(@connector.name.to_s)}_Mapper"
```
I've read through the code, but not yet understood how things fit together so please pardon my question:

Should we be using only names to identify mappers? You may have multiple ways to connect two models together, sometimes using different PK/FKs, and sometimes using the same PK/FK, but different conditions. In those cases would we always be using the same mapper? (this isn't a rhetorical question, I don't know the answer, but I thought it might be important to bring up)

dkubb commented on lib/data_mapper/mapper/attribute/embedded_value.rb:
```diff
@@ -26,7 +30,8 @@ def inspect
             mapper.load(tuple)
-          rescue
+          rescue => e
+            warn "AAA: #{e.message}"
```
Looks like a debug statement was left behind.

dkubb commented on lib/data_mapper/finalizer.rb:
```diff
@@ -1,13 +1,24 @@
 
     def initialize(mappers)
-      @mappers = mappers
+      @mappers           = mappers
```
That's alot of state in this constructor. fwiw, anytime I use more than 3 ivars in an object I stop to think if maybe there are objects or abstractions that are missing.

It's true that there are some cases where you do need more than 3 separate pieces of data in an object (like when constructing a SELECT statement that has 7-10 distinct parts), but I think it's an edge case.

dkubb commented:
@snusnu can you describe, maybe in the PR description, more of a high level of what the benefits and drawbacks are for this PR?

Also, I wonder if you might want to consider using `Adamantium`, or basic immutable objects for this code? I think it'll force you down a different path in a few places. (in a good way though) I'd say the same thing about metrics. I realize this is mostly a spike (as you said in IRC), but I think you'll probably result in a much different direction for the code. 

There was a time when I used to think that you could design an API for code at a high level, and then later on refactor all the code underneath to be nicer to maintain. I still think that's true for the thinnest layer that the end user is expected to interact with, but only to a point. After writing veritas and a few other gems with the same metrics feedback loop I believe you end up spending a large amount of time trying to support interfaces and layers that wouldn't have existed in the first place. I'm not saying I have a problem with the code, but there are definitely things that would've been identified early that you would've had to eliminate that would probably affect much of the code.

I think @mbj and @solnic know what I'm talking about in this regard. It's certainly challenging, and probably slower than other approaches, but I think the results are far beyond anything any of us has done before adopting it.

solnic commented:
@dkubb initially @snusnu implemented relation graph like you described, with nodes wrapping base relations, edges connecting nodes and connectors using info from edges to get source/target and join them. I guess initial implementation tried to solve too many problems at once so I felt like it became a dead end. So I tried to clean it up a bit but it was too hard to understand what's going on there. So I took a couple of steps back. I created a base graph class, make the registry inherit from that class which gave a nice reduction of code in the registry class. When I had basic graph functionality in place I added nodes that wrap base relations and edges that connect nodes and hold information about relationship and aliases. I didn't want to introduce separate connectors to avoid complicating the code. I got to a point where I managed to build the graph with all the base relations as well as joined relations and it works even with multi-through joins which is awesome.

So this was part I. Now it's time for part II. And here I agree with what @snusnu wrote - we now miss things that his initial implementation had. RelationConnector == Edge right now, we can simplify edges by re-adding Connector concept that receives edge and relationship and knows how to build a join from both sides of the edge. We also need to re-add information about which attributes are used for join.

There's another important change I made. Mappers no longer work directly with veritas relations. They now receive relation graph nodes which quack like relations. This makes possible to easily add support for other relational backends like Arel. We just need to come up with an Arel engine and wrap it up inside a relation node subclass. Some configuration abstraction is also needed so that we can setup both Vertas gateways or other things, like Arel tables or mongo driver collections. I'm super psyched about this to be honest.

Regarding metrics - I have set them up here but we didn't use them for this PR yet. So after some reconsideration I'm planning to work with @snusnu to re-add stuff that he created to the current implementation. We can go through metrics too and refactor smelly parts.

solnic committed:
Add RelationRegistry::Connector and rename RelationConnector to RelationEdge

solnic committed:
Mark pending specs for 1.8.7 :rage:

solnic committed:
Rename @_mappers to @mappers

solnic committed:
Add VeritasEngine encapsulating veritas-specific setup

solnic committed:
Move relation registry to the engine

solnic committed:
Allow failures on ree for now (also :punch: ruby 1.8.7)

solnic committed:
Ooops. Fix rspec version pin

solnic committed:
Freeze the relation graph after finalization

solnic committed:
Deep-freeze relation graph :joy:

solnic committed:
Actually call super in RelationRegistry#freeze

snusnu committed:
Add SpecHelper.draw_relation_registry to draw graphs

solnic committed:
Move veritas-specific relation logic to subclass of RelationNode

solnic committed:
Refactor Connector::Builder.call

solnic committed:
[WIP] Build correctly named nodes

  * Nodes from connectors should have correct names now
  * Rename and simplify RelationNode::VeritasRelation#clone_for => #clone
  * Simplify mapper registry by using relationship names instead of objects
  * Change Connector::Builder so it creates connector and node for further access
  * Add Connector#source_name which should be used to determine source aliases
    when building mapper classes
  * Connector no longer sets its own name, it receives it in the constructor

solnic committed:
Add spec for "tags have many songs through song_tags" (yes! :satisfied:)

solnic committed:
Oops

solnic committed:
And one more oops

snusnu committed:
Add spec for inverse intermediate relationship

Also remove explicitly given :target_key option as
it is inferred automatically.

solnic committed:
Remove drawing of graph from m:m-through spec

solnic committed:
[WIP] Break down connector builder into 2 classes

snusnu committed:
Test if inverse M:N uses the same join relation

solnic committed:
tmp

snusnu committed:
Simplify RelationRegistry::Connector#via?

snusnu committed:
Remove currently unused Graph#edges_for method

solnic committed:
[WIP] Prepare aliases up-front and get rid of aliasing on-the-fly

solnic committed:
[WIP] Refactor relation edge/node builder

solnic committed:
[WIP] Use Equalizer to identify mappers and relationships

solnic committed:
[WIP] Uncomment stuff

solnic committed:
[WIP] Add finalizer integration spec

solnic committed:
[WIP] Mark pending examples

solnic committed:
[WIP] Make most of the specs pass

solnic committed:
[WIP] Still wip but all the cool specs are passing...:scream: :star: :boom:

solnic committed:
[WIP] DRY up relation registry builder and add more finalizer specs

solnic committed:
[WIP] Further DRYup of relation registry builder

solnic committed:
[WIP] Use relationship names for edge names

solnic committed:
Add edge names to the graphviz visualization

solnic committed:
Add MapperRegistry#relation_map

solnic committed:
Mark a spec as pending for Ruby 1.8.7

solnic committed:
Include equalizer in relation node

solnic committed:
Pin veritas ref and remove veritas-optimizer (temporarly)

dkubb commented on lib/data_mapper/alias_set.rb:
```diff
@@ -0,0 +1,56 @@
+module DataMapper
+  class AliasSet
```
Could this object be updated to use `Adamanitum`? It looks like it's written in the same style as veritas, where state is not mutated outside of the `#initialize` method.

solnic replied:
Yes I'm planning to use adamantium in many classes. This one's included.

dkubb commented on lib/data_mapper/alias_set.rb:
```diff
@@ -0,0 +1,56 @@
+
+    def each(&block)
+      return to_enum unless block_given?
+      @index.each { |field, alias_name| yield(field, alias_name) }
```
What about something like `@index.each { |args| yield *args }`. It doesn't seem necessary to expose the detail here that there is a key/value for each index entry.

dkubb commented on lib/data_mapper/alias_set.rb:
```diff
@@ -0,0 +1,56 @@
+    private
+
+    def initialize_index(attributes)
+      @index = attributes.primitives.each_with_object({}) do |attribute, fields|
```
It seems like this doesn't use any object state inside `AliasSet`. Could this be feature envy? What about having a method inside `Mapper::AttributeSet` that does this work and returns the index?

dkubb replied:
Actually, I see where it uses `alias_for`, and thus `@prefix`. What about having a method on `Mapper::AttributeSet` where you pass in `@prefix` and it handles returning an index for you?

solnic replied:
Good point!

dkubb commented on lib/data_mapper/engine.rb:
```diff
@@ -0,0 +1,21 @@
+
+    # @api private
+    def base_relation(name, header)
+      raise NotImplementedError, "#{self.class}#base_relation must be implemented"
```
You probably don't have to hard code the method name here, you can just use `__method__` instead.

dkubb commented on lib/data_mapper/engine.rb:
```diff
@@ -0,0 +1,21 @@
+
+    # @api private
+    def gateway_relation(relation)
+      raise NotImplementedError, "#{self.class}#gateway_relation must be implemented"
```
You can use `__method__` here too. (and anywhere else you have something similar)

dkubb commented on lib/data_mapper/finalizer.rb:
```diff
@@ -20,16 +32,64 @@ def run
+      end
+
+      @base_relation_mappers.each do |mapper|
+        mapper.relations.freeze unless mapper.relations.frozen?
```
I would probably just call `#freeze` on the object without the test. If it is already frozen it's a noop and quickly exists anyway.

dkubb commented on lib/data_mapper/mapper/relation.rb:
```diff
@@ -60,7 +82,7 @@ def inspect
       def find(options)
         restriction = @relation.restrict(Query.new(options, @attributes))
-        self.class.new(restriction.optimize)
+        self.class.new(restriction)
```
Why was `#optimize` removed here? In general I would recommend calling `#optimize` on every relation before you store it inside another object's state. If the relation is optimized it will just return self, otherwise it will simplify it's internal structure; and that's probably what you want to store.

solnic replied:
I know and I will restore it when you upgrade optimizer to work with edge veritas :)

dkubb replied:
Right, I'll do that as soon as I finish the conversion. Some pending specs started to pass unexpectedly, so I need to figure out why before I push them ;)

solnic replied:
Thanks man. The reason we use edge veritas is mostly because it accepts a hash in #restrict :)

dkubb replied:
@solnic fwiw, I pushed to edge veritas-optimizer tonight. It should work with edge veritas and the adamantium conversion now.

solnic replied:
@dkubb sorry man but edge veritas stack with adamantium breaks DO. see my comment on adamantium's [issue](https://github.com/dkubb/adamantium/pull/4#issuecomment-9859387)

dkubb commented on lib/data_mapper/finalizer.rb:
```diff
@@ -1,15 +1,27 @@
+      @mappers         = mappers
+      @mapper_registry = Mapper.mapper_registry
+      @edge_builder    = RelationRegistry::Builder
+      @mapper_builder  = Mapper::Builder
```
I might recommend injecting these classes into your constructor instead of hard coding them here, eg:

``` ruby
def initialize(mappers, edge_builder = RelationRegistry::Builder, mapper_builder = Mapper::Builder)
  # ...
end
```

I mentioned this to @snusnu in irc with a bit more detail: http://irclogger.com/.datamapper/2012-10-27#1351379401-1351379924

solnic replied:
Yes I use dependency injection in cases like that too. Thanks for pointing this out.

dkubb commented on lib/data_mapper/mapper/builder.rb:
```diff
@@ -0,0 +1,72 @@
+          klass.send(:include, Relationship::OneToMany::Iterator)
+        end
+
+        klass.finalize_attributes
```
If this is a command method you could just have this be the last statement in the method rather than using a bare `klass`.

dkubb commented on lib/data_mapper/mapper/builder.rb:
```diff
@@ -0,0 +1,72 @@
+      end
+
+      def mapper_name
+        "#{@source_mapper.name}_X_#{Inflector.camelize(@connector.name.to_s)}_Mapper"
```
The `#to_s` method is called on anything interpolated within `#{...}` so you can probably remove it here.

dkubb commented on lib/data_mapper/relation_registry.rb:
```diff
@@ -1,27 +1,117 @@
+    #
+    # @api private
+    def [](name)
+      @nodes.detect { |node| node.name == name.to_sym }
```
What about calling `name = name.to_sym` outside the loop, and then comparing against `name` inside it? It seems unnecessary to do it every iteration.

solnic replied:
Ha! Nice catch :)

dkubb commented on lib/data_mapper/relation_registry/builder.rb:
```diff
@@ -0,0 +1,92 @@
+
+      # @api private
+      def build_node(name, relation)
+        unless relations[name]
```
What about:

``` ruby
def build_node(name, relation)
  relations[name] || begin
    node = relations.build_node(name, relation)
    relations.add_node(node)
    node
  end
end
```

solnic replied:
Actually I was planning to add `RelationRegistry#fetch_edge`  and `RelationRegistry#fetch_node` that would work like `Hash#fetch`. WDYT?

dkubb replied:
@solnic how do you see these being implemented, would `#fetch_node` just do `relations.fetch(name, &block)` or something similar?

solnic replied:
@dkubb `nodes.fetch(name, &block)` so I could write:

``` ruby
node = relations.fetch_node(name) {
  node = relations.build_node(name, relation)
  relations.add_node(node)
  node
}
```

dkubb commented on lib/data_mapper/relation_registry/builder.rb:
```diff
@@ -0,0 +1,92 @@
+
+      # @api private
+      def build_edge(name = relationship.name, left = left_node, right = right_node)
+        edge = relations.edge_for(left, right)
```
You could probably use the same thing here as I mentioned above in `build_node`.

dkubb commented on lib/data_mapper/relation_registry/builder/base_builder.rb:
```diff
@@ -0,0 +1,18 @@
+
+        # @api private
+        def name
+          @name ||= NodeName.new(left_name, right_name, relationship.name).to_connector_name
```
What about using `Adamanitum` and then using `memoize :name` here?

dkubb commented on lib/data_mapper/relation_registry/builder/node_name.rb:
```diff
@@ -0,0 +1,54 @@
+        end
+
+        # @api private
+        def to_str
```
The `#to_str` and `#to_ary` methods are generally supposed to be used with objects that are interface compatible with `String` and `Array`.

dkubb commented on lib/data_mapper/relation_registry/relation_node.rb:
```diff
@@ -0,0 +1,21 @@
+    class RelationNode < Graph::Node
+      include Enumerable
+
+      include Equalizer.new(:name, :relation)
```
Should the `aliases` be factored into the equalizer? The general pattern is that all the object's state is used.

dkubb commented on lib/data_mapper/relation_registry/relation_node/veritas_relation.rb:
```diff
@@ -0,0 +1,55 @@
+        end
+
+        def base?
+          veritas_relation = relation.respond_to?(:relation) ? relation.send(:relation) : relation
```
Is it necessary to use `#send` here if the object responds to `#relation`?

solnic replied:
Yeah I wanted to ask you about this. Gateway relation has private reader for relation hence the send.

dkubb replied:
Would it be better to have a base? method on the gateway?

solnic replied:
Yes it would be nice!

dkubb commented on lib/data_mapper/relationship.rb:
```diff
@@ -1,114 +1,40 @@
   #
   # @api private
   class Relationship
+    include Equalizer.new(:name, :source_model, :target_model)
```
It looks like there's a lot more state inside the object. Should this use more than just a few parts of it? I can't imagine you would want two relationships between the same models to be considered equivalent if they use different attributes to join on. (for example)

dkubb commented on lib/data_mapper/support/graph/edge.rb:
```diff
@@ -0,0 +1,20 @@
+      @name  = name
+      @left  = left
+      @right = right
+      @nodes = Set.new([ @left, @right ])
```
This is an _extremely_ tiny point, but this creates an `Array` unnecessarily. You can probably get the same result by doing `Set.new << @left << @right`.

If you do decide you don't care about the unnecessary object creation, a slightly more idiomatic way to write what you've done here is: `Set[ @left, @right ]`. It still creates an `Array` object (because it splats the arguments when passed in), but it basically passes that to `Set.new` the same as your code.

dkubb commented on lib/data_mapper/support/graph/node.rb:
```diff
@@ -0,0 +1,9 @@
+    attr_reader :name
+
+    def initialize(name)
+      @name = name
```
I'd recommend changing this to `@name = name.to_sym` and then remove `.to_sym` from all the places where you pass the name into the constructor or descendant constructor.

dkubb commented on lib/data_mapper/finalizer.rb:
```diff
@@ -20,16 +32,64 @@ def run
+    # @api private
+    def relationships_for_target(model)
+      @base_relation_mappers.map { |mapper|
+        relationships     = mapper.relationships.select { |relationship| relationship.target_model == model }
```
What about using a tighter matcher on this, like `#equal?` ?

dkubb commented on lib/data_mapper/finalizer.rb:
```diff
@@ -20,16 +32,64 @@ def run
 
     def finalize_attribute_mappers
-      @mappers.each { |mapper| mapper.finalize_attributes }
+      mappers.each { |mapper| mapper.finalize_attributes }
```
You could probably simplify this to `mappers.each(&finalize_attributes)`.

solnic committed:
Mark pending spec for ruby 1.8.7

solnic committed:
Simplify AliasSet#each

solnic committed:
Move responsibility of building an alias index to AttributeSet

solnic committed:
Minor simplification of Engine abstract methods

solnic committed:
Inject builder classes into Finalizer constructor

solnic committed:
Use tighter matcher to find relationships for given target models

solnic committed:
Simplify Finalizer#finalize_attribute_mappers

solnic committed:
No need to check relations.frozen?

solnic committed:
Minor simplification of Mapper::Builder#mapper_name

solnic committed:
Simplify RelationRegistry#[]

solnic committed:
Move #to_sym calls on name to Graph::Node and Graph::Edge

solnic committed:
Simplify set creation in Graph::Edge

solnic committed:
Use Equalizer in Graph::Edge

solnic committed:
Remove obsolete nasty hack

solnic committed:
Remove ree from "allow_failures"

solnic committed:
Adjust flay config

solnic commented:
@dkubb @snusnu I merged it in as the PR was too big already and I need a fresh start for further refactoring. I'm planning to make "the metrics happy" and improve test coverage. I hope bigger refactorings are no longer needed. I'd like to talk to you guys about the generated graph too so we can make sure it's correct.

The outcome of this refactoring is really awesome. Even if the graph still needs some work we introduced a pretty nice abstraction "by an accident" which is great. Once we're happy with the implementation I want to focus on ARel engine. It's not going to be as clean as in case of Veritas but it is doable.

I will open a new PR as soon as I have more refactoring done so we can discuss further.
